templates/boltzmann/middleware/honeycomb.ts

```diff
@@ -9,77 +19,180 @@ import beeline from 'honeycomb-beeline'
+            [SemanticAttributes.HTTP_METHOD]: context.method,
+            [SemanticAttributes.HTTP_SCHEME]: context.url.protocol,
+            [SemanticAttributes.HTTP_ROUTE]: context.url.pathname,
+            [OTEL_REQ_QUERY]: context.url.search
+          }
+        })
 
-      if (!trace) {
-        return next(context)
-      }
+        // this propagator takes care of extracting trace parent
+        // and state from request headers (and so on)
+        const propagator = new W3CTraceContextPropagator()
 
-      const boundFinisher = beeline.bindFunctionToTrace((response: ServerResponse) => {
-        beeline.addContext({
-          'response.status_code': String(response.statusCode)
-        })
+        propagator.inject(
+          otelTrace.setSpanContext(
+            ROOT_CONTEXT,
+            parentSpan.spanContext()
+          ),
+          carrier,
+          defaultTextMapSetter
+        )
 
-        beeline.addContext({
-          'request.route': context.handler.route,
-          'request.method': context.handler.method,
-          'request.version': context.handler.version
-        })
+        // create a parent active context
+        const parentContext = propagator.extract(
+          ROOT_CONTEXT,
+          carrier,
+          defaultTextMapGetter
+        )
 
-        const params = Object.entries(context.params).map(([key, value]) => {
-          return [`request.param.${key}`, value]
+        // set the active context
+        otelContext.with(parentContext, () => {
+          // keep the context active until we close the span
+          return new Promise((resolve, reject) => {
+            // do not do as I do,
+            onHeaders(context._response, function () {
+              // do our dangest to capture/handle surprise errors
+              try {
+                const handler: Handler = <Handler>context.handler
+
+                parentSpan.setAttribute(
+                  SemanticAttributes.HTTP_STATUS_CODE,
+                  String(context._response.statusCode)
+                )
+                parentSpan.setAttribute(
+                  SemanticAttributes.HTTP_ROUTE,
+                  <string>handler.route
+                )
+                parentSpan.setAttribute(
+                  SemanticAttributes.HTTP_METHOD,
+                  <string>handler.method
+                )
+                parentSpan.setAttribute(
+                  SemanticResourceAttributes.SERVICE_VERSION,
+                  <string>handler.version
+                )
+
+                Object.entries(context.params).map(([key, value]) => {
+                  parentSpan.setAttribute(
+                    `${OTEL_REQ_PARAM_NAMESPACE}.${key}`,
+                    value
+                  )
+                })
+                parentSpan.end()
+              } catch (err) {
+                // we don't want to crash the route just because
+                // otel didn't work...
```

Comment:
There's a general philosophy espoused by the otel docs of never crashing because of a snafu with telemetry, and it occurred to me that there *could* be an edge case here.